tests/matrix_free/matrix_vector_hessians_faces.cc

```diff
@@ -0,0 +1,388 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2013 - 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/logstream.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_q_generic.h>
+
+#include <deal.II/grid/grid_generator.h>
+
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/la_parallel_vector.h>
+
+#include <deal.II/matrix_free/fe_evaluation.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+#include "../tests.h"
+
+// tests if applying a matrix vector product with a matrix containing hessians
+// on faces produces the same result with FEFaceEvaluation and FEFaceValues
+// this is checked for different combinations of EvaluationFlags, FE types and
+// polynomial degrees
+
+template <int dim>
+void
+test_hessians(const unsigned int                             degree,
+              const dealii::FE_Poly<dim> &                   fe,
+              const dealii::EvaluationFlags::EvaluationFlags evaluation_flags)
+{
+  using namespace dealii;
+  using VectorizedArrayType = VectorizedArray<double>;
+
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria);
+  for (int i = 0; i < 3; ++i)
+    tria.refine_global();
```

Comment:
```suggestion
  tria.refine_global(3);
```
As a side question: Does the code also work if we have hanging nodes, i.e., the first cell or a few cells are refined adaptively? In that case it would make sense to add that here to make the test general.

Reply:
I tried it out and FEFaceValues and FEFaceEvaluation have different results in this case. Should we address this in this PR?

Reply:
Let's postpone this!

Reply:
Fine for me. I added a `TODO` to the test.